pallets/order-book/src/lib.rs

```diff
@@ -0,0 +1,662 @@
+			orders.retain(|o| *o != order.order_id);
+			<AssetPairOrders<T>>::insert(order.asset_in_id, order.asset_out_id, orders);
+			Ok(())
+		}
+
+		/// Generate a hash to be used for the order id using the placing
+		/// account, asset IDs, and pallet nonce.
+		pub fn gen_hash(
+			placer: &T::AccountId,
+			asset_out: T::AssetCurrencyId,
+			asset_in: T::AssetCurrencyId,
+			nonce: T::Nonce,
+		) -> T::Hash {
+			(&placer, asset_in, asset_out, nonce).using_encoded(T::Hashing::hash)
+		}
+
+		/// Get reserve amount when fee and out currency are the same
+		pub fn get_combined_reserve(
+			sell_amount: T::ForeignCurrencyBalance,
+		) -> Result<FeeBalance<T>, DispatchError> {
+			let fee_amount = T::Fees::fee_value(T::OrderFeeKey::get());
+
+			let sell_reserve_balance: FeeBalance<T> = sell_amount
+				.try_into()
+				.map_err(|_| Error::<T>::BalanceConversionErr)?;
+			Ok(sell_reserve_balance.ensure_add(fee_amount)?)
+		}
+	}
+
+	impl<T: Config> TokenSwaps<T::AccountId> for Pallet<T>
+	where
+		<T as frame_system::Config>::Hash: PartialEq<<T as frame_system::Config>::Hash>,
+	{
+		type Balance = T::ForeignCurrencyBalance;
+		type CurrencyId = T::AssetCurrencyId;
+		type OrderId = T::Hash;
+
+		/// Creates an order.
+		/// Verify funds available in, and reserve for  both chains fee currency
+		/// for storage fee, and amount of outgoing currency as determined by
+		/// the buy amount and price.
+		fn place_order(
+			account: T::AccountId,
+			currency_in: T::AssetCurrencyId,
+			currency_out: T::AssetCurrencyId,
+			buy_amount: T::ForeignCurrencyBalance,
+			sell_price_limit: T::ForeignCurrencyBalance,
+			min_fullfillment_amount: T::ForeignCurrencyBalance,
+		) -> Result<Self::OrderId, DispatchError> {
+			ensure!(currency_in != currency_out, Error::<T>::ConflictingAssetIds);
+
+			ensure!(
+				buy_amount != T::ForeignCurrencyBalance::zero(),
+				Error::<T>::InvalidBuyAmount
+			);
+			ensure!(
+				sell_price_limit != T::ForeignCurrencyBalance::zero(),
+				Error::<T>::InvalidMinPrice
+			);
+			ensure!(
+				T::AssetRegistry::metadata(&currency_in).is_some(),
+				Error::<T>::InvalidAssetId
+			);
+			ensure!(
+				T::AssetRegistry::metadata(&currency_out).is_some(),
+				Error::<T>::InvalidAssetId
+			);
+			<NonceStore<T>>::try_mutate(|n| {
+				*n = n.ensure_add(T::Nonce::one())?;
+				Ok::<_, DispatchError>(())
+			})?;
+			let max_sell_amount = buy_amount.ensure_mul(sell_price_limit)?;
+
+			let fee_amount = T::Fees::fee_value(T::OrderFeeKey::get());
+			if T::FeeCurrencyId::get() == currency_out {
+				let total_reserve_amount = Self::get_combined_reserve(max_sell_amount)?;
+				T::ReserveCurrency::reserve(&account, total_reserve_amount)?;
+			} else {
+				T::ReserveCurrency::reserve(&account, fee_amount)?;
+
+				T::TradeableAsset::reserve(currency_out, &account, max_sell_amount)?;
+			}
+
+			let new_nonce = <NonceStore<T>>::get();
+			let order_id = Self::gen_hash(&account, currency_in, currency_out, new_nonce);
```

Comment:
I do not understand why not just use `new_nonce` as an `order_id`. So instead of using a Hash type, use the nonce number directly as identification.

Reply:
Yeah, the nonce would work here, and simplify things a bit.  I might have been a bit paranoid with the order id determinability.

Reply:
There's only one way of identifying the order in the system right now, the nonce is just used to ensure each order gets a new hash, but I haven't thought of any ways that someone would be able to maliciously use a predicted order id, so I think it would be fine to change it to nonce.

Reply:
Updated to use nonce for order id.